// dgeni-packages/typescript/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* tslint:disable:only-arrow-functions */
const dgeni_1 = require("dgeni");
const linkInheritedDocs_1 = require("./processors/linkInheritedDocs");
const mergeParameterInfo_1 = require("./processors/mergeParameterInfo");
const readTypeScriptModules_1 = require("./processors/readTypeScriptModules");
const exportSymbolsToDocsMap_1 = require("./services/exportSymbolsToDocsMap");
const modules_1 = require("./services/modules");
const host_1 = require("./services/ts-host/host");
const TsParser_1 = require("./services/TsParser");
// Define the dgeni package for generating the docs
module.exports = new dgeni_1.Package('typescript', [require('../jsdoc')])
    // Register the services and file readers
    .factory(modules_1.modules)
    .factory(exportSymbolsToDocsMap_1.exportSymbolsToDocsMap)
    .factory('tsHost', function tsHost() { return new host_1.Host(); })
    .factory('tsParser', function (log) { return new TsParser_1.TsParser(log); })
    .factory('convertPrivateClassesToInterfaces', function convertPrivateClassesToInterfaces() { return convertPrivateClassesToInterfaces; })
    .factory('EXPORT_DOC_TYPES', function EXPORT_DOC_TYPES() {
    return [
        'class',
        'interface',
        'function',
        'var',
        'const',
        'let',
        'enum',
        'type-alias',
        'value-module',
    ];
})
    // Register the processors
    .processor(readTypeScriptModules_1.readTypeScriptModules)
    .processor(linkInheritedDocs_1.linkInheritedDocs)
    .processor(mergeParameterInfo_1.mergeParameterInfo)
    // Configure ids and paths
    .config(function (computeIdsProcessor, computePathsProcessor, EXPORT_DOC_TYPES) {
    computeIdsProcessor.idTemplates.push({
        docTypes: ['member'],
        idTemplate: '${containerDoc.id}.${name}',
        getAliases(doc) {
            return doc.containerDoc.aliases.map((alias) => alias + '.' + doc.name);
        },
    });
    computePathsProcessor.pathTemplates.push({
        docTypes: ['member'],
        pathTemplate: '${containerDoc.path}#${name}',
        getOutputPath() {
            // These docs are not written to their own file, instead they are part of their class doc
        },
    });
    computePathsProcessor.pathTemplates.push({
        docTypes: ['parameter'],
        pathTemplate: '${container.path}#${name}',
        getOutputPath() {
            // These docs are not written to their own file, instead they are part of their callable doc
        },
    });
    const MODULES_DOCS_PATH = 'partials/modules';
    computePathsProcessor.pathTemplates.push({
        docTypes: ['module'],
        outputPathTemplate: MODULES_DOCS_PATH + '/${id}/index.html',
        pathTemplate: '/${id}',
    });
    computePathsProcessor.pathTemplates.push({
        docTypes: EXPORT_DOC_TYPES,
        outputPathTemplate: MODULES_DOCS_PATH + '/${path}/index.html',
        pathTemplate: '${moduleDoc.path}/${name}',
    });
});
